src/core/DY.Core.Label.Type.fst

```diff
@@ -1,93 +1,35 @@
-/// it is in practice behaving like `join (S prin 0) (join (S prin 1) (join ...))`.
-/// It is hard-coded like this because infinite joins are not supported:
-/// only finite joins are supported (via the binary join that can be folded on a list).
-
-type pre_label =
-  | P: principal -> pre_label
-  | S: principal -> state_id -> pre_label
-
-/// Labels are roughly a free lattice on pre-labels,
-/// with lower bound (meet) and upper bound (join),
-/// as well as a minima (secret) and maxima (public).
-
-type label =
-  | Secret: label
-  | State: pre_label -> label
-  | Meet: label -> label -> label
-  | Join: label -> label -> label
-  | Public: label
-
-open DY.Core.Internal.Ord
-
-instance integer_encodable_pre_label: integer_encodable pre_label = {
-  encode = (fun x ->
-    match x with
-    | P p -> 0::(encode p)
-    | S p s -> 1::(encode [encode p; encode s.the_id])
-  );
-  encode_inj = (fun x y ->
-    encode_inj_forall (list (list int)) ();
-    encode_inj_forall string ();
-    encode_inj_forall nat ()
-  );
-}
-
-val encode_label: label -> list int
-let rec encode_label l =
-  match l with
-  | Secret -> 0::[]
-  | State s -> 1::(encode [encode s])
-  | Meet l1 l2 -> 2::(encode [encode_label l1; encode_label l2])
-  | Join l1 l2 -> 3::(encode [encode_label l1; encode_label l2])
-  | Public -> 4::[]
-
-val encode_label_inj:
-  l1:label -> l2:label ->
-  Lemma
-  (requires encode_label l1 == encode_label l2)
-  (ensures l1 == l2)
-let rec encode_label_inj l1 l2 =
-  introduce forall l1' l2'. l1' << l1 /\ l2' << l2 /\ encode_label l1' == encode_label l2' ==> l1' == l2' with (
-    introduce _ ==> _ with _. (
-      encode_label_inj l1' l2'
-    )
-  );
-  encode_inj_forall (list (list int)) ();
-  encode_inj_forall pre_label ();
-  allow_inversion label
-
-instance integer_encodable_label: integer_encodable label = {
-  encode = encode_label;
-  encode_inj = encode_label_inj;
+/// We therefore want labels to be predicates `trace -> prop`,
+/// however this doesn't work with F*'s positivity checker.
+/// (see https://fstar-lang.org/tutorial/book/part2/part2_inductive_type_families.html#strictly-positive-definitions )
+/// Indeed, we want the trace to contain labels (see `RandGen` event),
+/// and we want labels to be predicates on the trace, namely `trace -> prop` (see DY.Core.Label.Type).
+/// A type `t` cannot contain `t -> prop` because this could be used to derive False (using Cantor's diagonal argument).
+/// To circumvent that, labels are predicates on traces with no labels (i.e. whose label type is `unit`).
+/// This solves the positivity issue because `trace_ label` contain labels that are `trace_ unit -> prop`,
+/// which respects the positivity condition.
+/// It means that labels cannot decide about their corruption depending on other labels in the trace,
+/// but this is not a problem in practice.
+
+[@@erasable]
+noeq
+type label = {
+  is_corrupt: trace_ unit -> prop;
+  is_corrupt_later:
```

Comment:
I recall a previous version used two traces `tr1 <$ tr2` here, and then had a constructor that only used the single-extension version. My intuition is that that is more convenient to use by hiding the induction principle, but it seems like this is now accomplished by an SMTPat on the lemma `is_corrupt_later`.

Was there any particular reason to prefer one over the other? The main reason I can see to prefer this version is that it's maybe more intuitive when creating a label, but I also don't see users creating labels manually very often, realistically.

Reply:
Indeed, it was changed in this commit: https://github.com/REPROSEC/dolev-yao-star-extrinsic/pull/55/commits/c672ce9d2ba86747a3410cab5103dd06a25d0d37#diff-ca8bc1f64c5cdb602875af699132c2ab7bde4f8e7e11481c7ec2dd5177d8484a

I would prefer to use `tr1 <$ tr2`, but I changed that to improve the dependency graph, which is a bit nasty because labels are predicates on the trace, and the trace contains labels, so it seems like we have a dependency cycle.

The module are defined in this order:
- `DY.Core.Bytes.Type` defines the `bytes` and `usage` type, and doesn't depend on other modules
- `DY.Core.Trace.Type` defines the `trace_` type which is parametrized by the label type. It only depends on `DY.Core.Bytes.Type`
- `DY.Core.Label.Type` defines the `label` type using `trace_` defined in `DY.Core.Trace.Type`
- `DY.Core.Trace.Base` defines the `trace` type which is just an alias for `trace_ label`, hence depending on the modules `DY.Core.Trace.Type` and `DY.Core.Label.Type`

The reason I switch from `tr1 <$ tr2` to `Snoc tr ev` is that otherwise, `DY.Core.Label.Type` would need the definition of `<$`, which would need to be in `DY.Core.Trace.Type`, and then it would be weird that some predicates / relations on traces are in `DY.Core.Trace.Base` and others in `DY.Core.Trace.Type`.
By using this new formulation, `DY.Core.Trace.Type` only defines the `trace_` type and nothing else.

In #56, I introduce a helper constructor for labels https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/d8dcf286e25f102695c0a716beb270427e4ce8af/src/core/DY.Core.Label.fst#L62-L75
Which mean that we can use `tr1 <$ tr2` to define labels, and do not have to deal with the weird property on `Snoc`.